d4s/src/main/scala/d4s/codecs/D4SDecoder.scala

```diff
@@ -73,12 +73,21 @@ object D4SDecoder {
       if (item.m().isEmpty) {
         ctx.subtypes.toList
           .find(_.typeName.short.contains(item.s()))
-          .toRight(new CannotDecodeAttributeValue(s" Cannot decode item of type ${ctx.typeName}.", None))
+          .toRight(new CannotDecodeAttributeValue(s" Cannot decode item of type ${ctx.typeName.full} from string: ${item.s()}", None))
           .flatMap(_.typeclass.decodeAttribute(item))
       } else {
+        // Looks awful ?
+        val typeName = item.m().asScala.keySet.head
         ctx.subtypes.toList
-          .collectFirstSome(_.typeclass.decode(item.m()).toOption)
-          .toRight(new CannotDecodeAttributeValue(s"Cannot decode item of type ${ctx.typeName}.", None))
+          .find(_.typeName.short.contains(typeName))
+          .toRight(new CannotDecodeAttributeValue(s"Cannot decode item of type ${ctx.typeName.full} from map: ${item.m().asScala.toMap}", None))
+          .flatMap {
+            subtype =>
+              item.m().asScala.get(typeName) match {
+                case None => Left(new CannotDecodeAttributeValue("SOME MEANINGFUL ERROR", None))
+                case Some(value) => subtype.typeclass.decodeAttribute(value)
+              }
+          }
       }
```

Comment:
@neko-kai, what do u think about that?

Reply:
You should check that map size is 1, then just get `.head` when the single existing key matches some typeName. You can use `.find` without `.toList`. `.contains(typeName)` is total non-sense, the key should be _equal_ to typeName